doc/manual/metaprogramming.rst

```diff
@@ -872,3 +872,102 @@ entirely in Julia. You can read their source and see precisely what they
 do — and all they do is construct expression objects to be inserted into
 your program's syntax tree.
 
+Staged functions
+----------------
+
+Sometimes, generating or manipulating expressions through macros is a little
+overkill for the problem you're trying to solve, but you still need some
+more advanced magic than just regular functions; enter *staged functions*.
+Staged functions have the capability to generate specialized code depending
+on the *types* of the arguments you give them, so that you can optimize or
+generalize your code in ways that aren't possible with ordinary functions.
```

Comment:
I think everything is there now:

Stagedfunctions play a similar role as macros, but at a later stage between parsing and run-time. Staged functions give the capability to generate specialized code depending on the *types* of their arguments. Macros work with expressions at parsing-time and cannot access the types of their inputs. In contrast, a stagedfunction gets expanded at a time when the type of the arguments is known, but the function is not yet compiled.
Depending on the types of the input, a staged function returns a *quoted expression* which then forms the function body of the specialized function. Staged functions provide a flexible framework to move work from runtime to compile-time.